### PR TITLE
drivers: can: Rework CAN drivers and API

### PR DESCRIPTION
**Pull request by alexanderwachter**
This PR extends the isr callback and the attach_isr function with an argument.
With this argument it's possible to map different filters to the same isr but having the possibility to distinguish them by the argument. This makes the implementation if higher level protocols easier but breaks the current API.
The sram overhead is sizeof(void*) * CAN_MAX_FILTER

EDIT:
attach_msgq is moved to a common function to all drivers. This reduces the complexity of the specific implementations.

attach_workq is implemented. This is useful to offload work from an ISR to a workq thread.
This function can be used to attach a callback even from userspace.
This is not possible for attach_isr.

### CODE REVIEW

**Comment by alexanderwachter** [comment]
@lowlander what do you think?

**Comment by alexanderwachter** [comment]
@karstenkoenig ping

**Review by jukkar — APPROVED** [review]
This makes lot of sense so +1

**Comment by lowlander** [comment]
My old driver that I have almost cleaned up enough to create a PR for, has a bit of another approach. I give my driver a (const) array of filters, that are directly used by the driver, so I do not copy them to not need writable RAM (or a fixed CONFIG_FILTER_COUNT option). Also I made it possible to set a filter-array per FIFO (there are two RX fifos on the F4) Also I have no "detach" system-call because just registering a new array (which can be 0) releases the previous one.  Those array entries, have not just the filter and callback, but also a user-pointer, that will be passed to the callback. 

Currently I am implementing a E70 CAN-FD driver, to make sure the API's work on more than just a STM32 controller.   

I am a bit afraid things are done in parallel at the moment, but last time I early asked for comments I was pretty much shot down due to code quality, so now I just finish my driver and create a PR when I am done.

**Comment by jukkar** [comment]
Regarding the API break that this PR is introducing, should we just keep the old API and create a new function that introduces the extra arg to the callback. The old API could be deprecated if needed and eventually removed at some later time.

**Comment by alexanderwachter** [comment]
@jukkar I'm not sure if the attach isr is used by users anyway. From userspace it's better to use the queue. The attach_isr is more useful for highlevel stuff. But making a wrapper will not produce any overhead. I'll do that.

**Comment by alexanderwachter** [comment]
> I give my driver a (const) array of filters, that are directly used by the driver, so I do not copy them to not need writable RAM (or a fixed CONFIG_FILTER_COUNT option)
I don't copy the filters neither. I just set the registers.  CONFIG_FILTER_COUNT is for the amount of different callbacks/message queues and they can't be constant because socket can relies on appending filter at run-time.
Also my ISO-TP stack I'm working on relies in attach/detach at run-time. Additionally an other project (the project I implement all the CAN stuff here) relies on attach and detach at run-time.

**Comment by alexanderwachter** [comment]
Added an argument to the send callback too.

**Comment by alexanderwachter** [comment]
Rebased.

**Comment by alexanderwachter** [comment]
can_attach_workq is implemented as a replacement for can_attach_isr when userspace is on.
attaching an ISR from userspace is not allowed.(See #14204)

**Review by jukkar — COMMENTED** [review]
One possible issue with the API changes is that we are changing the API here. There is the doc for API mgmt https://docs.zephyrproject.org/latest/development_process/api_lifecycle.html. For 1.14 lot of API changes were done that did not follow the rules verbatim, but it looks like that for 1.15 the rules are more strictly followed.
Anyway, as I am not really using the CAN low level APIs, perhaps those one that do could comment these changes.

**Comment by alexanderwachter** [comment]
@karstenkoenig can you check the changes for mcp2515?

File: drivers/can/can_loopback.c
Comment (rodrigopex) [inline drivers/can/can_loopback.c]: return -EIO...

File: drivers/can/can_loopback.c
Comment (rodrigopex) [inline drivers/can/can_loopback.c]: I would suggest using a standard error number. Something like -EBUSY. I am guessing there is no negative filter ID right?

File: drivers/can/can_loopback.c
Comment (rodrigopex) [inline drivers/can/can_loopback.c]: I usually put the unused variables at the first line of the function implementation. First the unused and after that, the code. I don't know if it is a good practice or just my of doing that.

File: drivers/can/can_loopback.c
Comment (rodrigopex) [inline drivers/can/can_loopback.c]: The same about the scope of the **i**. I always try to shrink that.

File: drivers/can/can_loopback.c
Comment (rodrigopex) [inline drivers/can/can_loopback.c]: Did you mean **{0}**?

File: samples/drivers/CAN/src/main.c
Comment (rodrigopex) [inline samples/drivers/CAN/src/main.c]: I usually change the position to ```u16_t cnt = (msg->data[1] << 8) | msg->data[0];``` It seems to be more natural.

File: samples/drivers/CAN/src/main.c
Comment (rodrigopex) [inline samples/drivers/CAN/src/main.c]: Same about UNUSED first...

**Review by rodrigopex — COMMENTED** [review]
Here are my 2 cents. I hope I could help here.

File: drivers/can/can_loopback.c
Comment (alexanderwachter) [reply drivers/can/can_loopback.c]: Right. But the CAN error numbers are defined along with the API to get a more precise error message.

File: drivers/can/can_loopback.c
Comment (alexanderwachter) [reply drivers/can/can_loopback.c]: I thought the definitions have to be at the beginning.

File: drivers/can/can_loopback.c
Comment (alexanderwachter) [reply drivers/can/can_loopback.c]: Right. I thought it is now allowed. I'll change it.

File: drivers/can/can_loopback.c
Comment (alexanderwachter) [reply drivers/can/can_loopback.c]: can_loopback_cfg_1 is not used at the moment.

File: drivers/can/CMakeLists.txt
Comment (alexanderwachter) [reply drivers/can/CMakeLists.txt]: That's right. I can make a commit for that.

File: drivers/can/can_common.c
Comment (alexanderwachter) [reply drivers/can/can_common.c]: Good point. I could also use void and throw the error log here.

Comment (alexanderwachter) [reply drivers/can/can_common.c]: -EIO seems not really to be ideal here because it is not an IO error. I can't find a suitable error number like buffer full. Since this is defined static and only used at one location I leave it as -1 here

File: drivers/can/CMakeLists.txt
Comment (alexanderwachter) [reply drivers/can/CMakeLists.txt]: Done

File: samples/drivers/CAN/src/main.c
Comment (alexanderwachter) [reply samples/drivers/CAN/src/main.c]: The sample needs to be updated anyways. I'll provide a new one after this gets merged.

File: drivers/can/can_loopback.c
Comment (alexanderwachter) [reply drivers/can/can_loopback.c]: Removed it.

File: drivers/can/can_handlers.c
Comment (alexanderwachter) [inline drivers/can/can_handlers.c]: This is obviously wrong. But which K_OBJ_ is the right one here? I can't find K_OBJ_WORKQ @andrewboie

Comment (andrewboie) [reply drivers/can/can_handlers.c]: a k_work_q is just a k_queue and a k_thread. we don't do anything to the thread, we are just going to be adding items to the k_queue object. so it's enough to do Z_OOPS(Z_SYSCALL_OBJ(&work_q->queue, K_OBJ_QUEUE).

File: drivers/can/can_handlers.c
Comment (andrewboie) [inline drivers/can/can_handlers.c]: No validation has been done to the work parameter. Have to prove that the caller has read/write access to it.

File: drivers/can/can_common.c
Comment (andrewboie) [inline drivers/can/can_common.c]: We have a security problem here.
The pointer to the zcan_work came from user memory. The problem is we will be using this data structure when the ISR fires. But what is going to stop the user thread from concurrently overwriting the members of the zcan_work struct with garbage or malicious data? If that happens, terrible things might happen can_work_isr_put() actually runs depending on what was changed, particularly the work->work_queue pointer.

**Review by tyrantlizard77 — DISMISSED** [review]
I'm seeing some problems with the workq integration with respect to user mode.
Doing this right may require some adjustments to the workq interfaces, I will follow up with some more thoughts on this.

File: drivers/can/can_common.c
Comment (alexanderwachter) [reply drivers/can/can_common.c]: How do I know if the work_q thread is in user mode?

File: drivers/can/can_common.c
Comment (alexanderwachter) [reply drivers/can/can_common.c]: @andrewboie I could check the data within the ISR. The user can't run concurrently to an ISR.

Comment (alexanderwachter) [reply drivers/can/can_common.c]: An alternative would be making the zcan_work a kernel object. But how can I copy the zcan_frame to the userspace then?

**Review by karstenkoenig — APPROVED** [review]
Looks all good to me, glad to see attachmsqq gone from the hardware drivers!
Also still works on my setup here :-)

File: drivers/can/Kconfig.loopback
Comment (alexanderwachter) [reply drivers/can/Kconfig.loopback]: Thanks! Resolved

File: drivers/can/can_mcp2515.c
Comment (alexanderwachter) [reply drivers/can/can_mcp2515.c]: They don't need to be initialized. They are not checked or used before an ISR is actually attached.

File: drivers/can/can_common.c
Comment (alexanderwachter) [reply drivers/can/can_common.c]: @andrewboie Userspace handler for attaching workqueues is removed until workqueues get proper userspace support.

Comment (andrewboie) [reply drivers/can/can_common.c]: since there is no handler can you remove the __syscall

Comment (alexanderwachter) [reply drivers/can/can_common.c]: Removed.

**Pull request merged**
